### PR TITLE
feat(shared-sub): add round_robin_per_group strategy

### PR DESCRIPTION
**Pull request by fridayy**
add initial implementation of a round robin strategy that balances load in a more predictable fashion

https://github.com/emqx/emqx/issues/8454

### CODE REVIEW

File: apps/emqx/i18n/emqx_schema_i18n.conf [hunk 1]
```diff
@@ -1124,6 +1125,7 @@ subscriber was not found, send to a random subscriber cluster-wide
             cn: """共享订阅的分发策略名称。
 - `random`: 随机选择一个组内成员；
 - `round_robin`: 循环选择下一个成员；
+- `round_robin_per_node`: ???；
```
Comment (fridayy) [inline apps/emqx/i18n/emqx_schema_i18n.conf:1128]: guess I would need some help here

Comment (lafirest) [reply apps/emqx/i18n/emqx_schema_i18n.conf:1128]: I think `round_robin_per_group` may be better,
round_roubin_per_group: 在共享组内循环选择下一个成员

File: apps/emqx/src/emqx_shared_sub.erl [hunk 2]
```diff
@@ -315,7 +317,10 @@ do_pick_subscriber(Group, Topic, round_robin, _ClientId, _SourceTopic, Count) ->
             N -> (N + 1) rem Count
         end,
     _ = erlang:put({shared_sub_round_robin, Group, Topic}, Rem),
-    Rem + 1.
+    Rem + 1;
+do_pick_subscriber(Group, Topic, round_robin_per_node, _ClientId, _SourceTopic, Count) ->
+    %% reset the counter back to zero if counter > subscriber count to avoid the counter grow larger than the current subscriber count
+    ets:update_counter(?SHARED_SUBS_ROUND_ROBIN_COUNTER, {Group, Topic}, {2, 1, Count, 1}).
```
Comment (fridayy) [inline apps/emqx/src/emqx_shared_sub.erl:323]: if configuration is changed at runtime - lets say from round_robin to round_robin_per_node the process will crash here. Is this something that must be handled or is changing configuration at runtime not a use-case?

File: apps/emqx/src/emqx_shared_sub.erl [hunk 3]
```diff
@@ -330,6 +335,7 @@ init([]) ->
     {atomic, PMon} = mria:transaction(?SHARED_SUB_SHARD, fun init_monitors/0),
     ok = emqx_tables:new(?SHARED_SUBS, [protected, bag]),
     ok = emqx_tables:new(?ALIVE_SUBS, [protected, set, {read_concurrency, true}]),
+    ok = emqx_tables:new(?SHARED_SUBS_ROUND_ROBIN_COUNTER, [public, set, {write_concurrency, true}]),
```
[the remark this answers is not in the document]
Comment (lafirest) [reply apps/emqx/src/emqx_shared_sub.erl:338]: I think this is fine

File: apps/emqx/src/emqx_shared_sub.erl (hunk 2 above)
Comment (lafirest) [reply apps/emqx/src/emqx_shared_sub.erl:323]: Hot reloading of the configuration is necessary. You should use ets:update_counter/4 here

**Comment by zmstone** [comment]
Hi @fridayy 
Thank you for the PR.
There are a few things to be fixed before this can be merged

* The last commit message has a space between `feat` and `(`, the script is not happy with it.
* Some lines are too long, the code style check is not passing, see the "Code style check / code_style_check" job for more details
* The test case emqx_shared_sub_SUITE:t_round_robin_per_group_two_nodes_publish_to_same_node is failing.

You can run this test suite locally by `./rebar3 ct -v --suite apps/emqx/test/emqx_shared_sub_SUITE.erl`

**Comment by fridayy** [comment]
> Hi @fridayy Thank you for the PR. There are a few things to be fixed before this can be merged
> 
> * The last commit message has a space between `feat` and `(`, the script is not happy with it.
> * Some lines are too long, the code style check is not passing, see the "Code style check / code_style_check" job for more details
> * The test case emqx_shared_sub_SUITE:t_round_robin_per_group_two_nodes_publish_to_same_node is failing.
> 
> You can run this test suite locally by `./rebar3 ct -v --suite apps/emqx/test/emqx_shared_sub_SUITE.erl`

hi @zmstone 
thanks for the hints, I will fix the issues asap

File: apps/emqx/src/emqx_shared_sub.erl [hunk 4]
```diff
@@ -395,6 +406,25 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%--------------------------------------------------------------------
 
+maybe_insert_round_robin_count({Group, _Topic} = GroupTopic) ->
+    strategy(Group) == round_robin_per_group andalso
```
Comment (zmstone) [inline apps/emqx/src/emqx_shared_sub.erl:410]:
```suggestion
    strategy(Group) =:= round_robin_per_group andalso
```

File: apps/emqx/src/emqx_shared_sub.erl [hunk 5]
```diff
@@ -395,6 +406,25 @@ code_change(_OldVsn, State, _Extra) ->
...
+        ets:insert(?SHARED_SUBS_ROUND_ROBIN_COUNTER, {GroupTopic, 0}),
+    ok.
+
+maybe_delete_round_robin_count({Group, _Topic} = GroupTopic) ->
+    strategy(Group) == round_robin_per_group andalso
```
Comment (zmstone) [inline apps/emqx/src/emqx_shared_sub.erl:415]:
```suggestion
    strategy(Group) =:= round_robin_per_group andalso
```

**Comment by zmstone** [comment]
Hi @fridayy 
Again, sorry for the delayed review. Just got back from vacation.
Could you please help to rebase your branch? 
also the last commit message has a space between `feat` and `(`, making the commit log style linter to fail. (maybe squash to the first commit).

**Comment by fridayy** [comment]
> Hi @fridayy Again, sorry for the delayed review. Just got back from vacation. Could you please help to rebase your branch? also the last commit message has a space between `feat` and `(`, making the commit log style linter to fail. (maybe squash to the first commit).

hi @zmstone - no worries I hope you had a good one! 
yeah sure. I will squash the commits hopefully not displease the linter again. ;)

**Comment by fridayy** [comment]
okay apparently  `t_round_robin_per_group_two_nodes_alternating_publish` is a flaky test - apologies. Will have a second look.

**Comment by fridayy** [comment]
Decided to remove both tests that involved another node. On a second thought they provide little to no value but only increase the test set execution time (by relying on sleep to help with ordering - which is a red flag itself) and have a high potential of being flaky when run on CI.

**Review by zmstone — APPROVED** [review]

**Comment by zmstone** [comment]
Thank you for the PR @fridayy

**Comment by fridayy** [comment]
> Thank you for the PR @fridayy

@zmstone my pleasure - thanks for the great support. have a nice day

**Pull request merged**